### PR TITLE
UPDATE for Vue3 ( = vuex 4.x, vue-router 4.x)

### PR DESCRIPTION
**Pull request by Spice-Z**
To use vuex-router-sync in Vue 3.x, I updated vue, vuex and vue-router!
Changed src and test, ~~currently the test is not perfect, help...😭~~

### ~~About vue-routers's afterEach~~
- ~~To mutate latest route info, this package using vue-router's `afterEach` function.~~
- ~~But vue-router v4.x [only export add function](https://github.com/vuejs/vue-router-next/blob/7b27e797a82d1d6a57a1f16bf43c43be3effb2cf/src/router.ts#L1057), and  doesn't export any list that registered afterEach.~~
- ~~I want to test `unsync` correctly  remove router's afterEach hook, but I have no idea.~~

### ~~About vuex's watch~~
- ~~To sync router on store change, this package using vuex's  `watch` function~~
- ~~But I can't find a way to confirm what is registered in watch hooks!~~
- ~~I want to test `unsync` correctly remove vuex's watch hook, but I have no idea...~~

### About unsync test
- Old unsync tests used some functions of vuex and vue-router.
- But in Vue 3.x, I cannnot use those functinos.
- I set new tests, directly watch store's state.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -44,7 +44,7 @@ export function sync(
   // sync router on store change
   const storeUnwatch = store.watch(
```
Comment (Danktuary) [inline src/index.ts:39]: The `{ sync: true }` below should be changed to `{ flush: 'sync' }` ([Vue 2](https://github.com/vuejs/vue/blob/b51430f598b354ed60851bb62885539bd25de3d8/src/core/observer/watcher.js#L34) vs [Vue 3](https://github.com/vuejs/vue-next/blob/master/packages/runtime-core/src/apiWatch.ts#L58))

File: package.json [hunk 2]
```diff
@@ -36,8 +36,8 @@
   },
   "homepage": "https://github.com/vuejs/vuex-router-sync#readme",
   "peerDependencies": {
-    "vue-router": "^3.0.0",
-    "vuex": "^3.0.0"
+    "vue-router": "^4.0.0-rc.5",
```
Comment (Akryum) [inline package.json]: Maybe use stable version range to keep it clean?

File: package.json [hunk 3]
```diff
@@ -59,8 +59,8 @@
     "ts-jest": "^25.4.0",
     "tslib": "^1.11.1",
     "typescript": "^3.8.3",
-    "vue": "^2.5.0",
-    "vue-router": "^3.0.0",
-    "vuex": "^3.0.0"
+    "vue": "^3.0.5",
+    "vue-router": "^4.0.2",
+    "vuex": "^4.0.0-rc.2"
```
Comment (Spice-Z) [inline package.json:64]: I want to update other dependencies, but that should be other PR to keep it simple...

File: src/index.ts (hunk 1 above)
Comment (Spice-Z) [reply src/index.ts:39]: Thank you! 
I fixed it!

File: src/index.ts [hunk 4]
```diff
@@ -1,36 +1,36 @@
 import { Store } from 'vuex'
-import VueRouter, { Route } from 'vue-router'
+import { Router, RouteLocationNormalized } from 'vue-router'
 
 export interface SyncOptions {
   moduleName: string
 }
 
 export interface State {
```
Comment (posva) [inline src/index.ts]: You should be able to do something like

```ts
export interface State extends Omit<RouteLocationNormalized, 'matched' | 'redirectedFrom'>	 {
}
```

and remove all the manal types

File: test/index.spec.ts [hunk 5]
```diff
@@ -1,54 +1,65 @@
-import Vue from 'vue'
-import Vuex, { mapState } from 'vuex'
-import VueRouter from 'vue-router'
-import { sync } from '@/index'
+import { createApp, defineComponent, h, computed, nextTick } from 'vue'
 
-Vue.use(Vuex)
-Vue.use(VueRouter)
+import { createStore, useStore } from 'vuex'
+import { createRouter, createWebHistory, RouterView } from 'vue-router'
```
Comment (posva) [inline test/index.spec.ts]: Since tests are running on Node, it's safer to use `createMemoryHistory`

File: test/index.spec.ts [hunk 6]
```diff
@@ -1,54 +1,65 @@
...
     foo: 'a',
     bar: 'b'
   })
 
-  const app = new Vue({
-    store,
-    router,
-    render: (h) => h('router-view')
-  }).$mount()
+  const rootEl = document.createElement('div')
```
Comment (posva) [inline test/index.spec.ts:50]: Wouldn't it be easier and safer to use vue test utils?

**Review by posva — CHANGES_REQUESTED** [review]

File: test/index.spec.ts (hunk 6 above)
Comment (Spice-Z) [reply test/index.spec.ts:50]: Ok, I will try to use vue test utils!
(Actually, I have not use it yet 😅 )

**Comment by kiaking** [comment]
Thank you so much for the PR! I've added test for the time traveling and now the coverage is 100% 🙌 

As @posva mentioned, I think we should use vue-test-utils ideally but since it's quite small test, I would go ahead and merge this one, then maybe add playground to do the final manual test, then I think we're ready to go 👍

**Review by kiaking — APPROVED** [review]

**Pull request merged**